In [2]:
import nltk

## Analizando Moby Dick

In [1]:
from nltk.book import *
text1

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


<Text: Moby Dick by Herman Melville 1851>

In [6]:
len(text1.tokens)

260819

In [3]:
text1.tokens

['[',
 'Moby',
 'Dick',
 'by',
 'Herman',
 'Melville',
 '1851',
 ']',
 'ETYMOLOGY',
 '.',
 '(',
 'Supplied',
 'by',
 'a',
 'Late',
 'Consumptive',
 'Usher',
 'to',
 'a',
 'Grammar',
 'School',
 ')',
 'The',
 'pale',
 'Usher',
 '--',
 'threadbare',
 'in',
 'coat',
 ',',
 'heart',
 ',',
 'body',
 ',',
 'and',
 'brain',
 ';',
 'I',
 'see',
 'him',
 'now',
 '.',
 'He',
 'was',
 'ever',
 'dusting',
 'his',
 'old',
 'lexicons',
 'and',
 'grammars',
 ',',
 'with',
 'a',
 'queer',
 'handkerchief',
 ',',
 'mockingly',
 'embellished',
 'with',
 'all',
 'the',
 'gay',
 'flags',
 'of',
 'all',
 'the',
 'known',
 'nations',
 'of',
 'the',
 'world',
 '.',
 'He',
 'loved',
 'to',
 'dust',
 'his',
 'old',
 'grammars',
 ';',
 'it',
 'somehow',
 'mildly',
 'reminded',
 'him',
 'of',
 'his',
 'mortality',
 '.',
 '"',
 'While',
 'you',
 'take',
 'in',
 'hand',
 'to',
 'school',
 'others',
 ',',
 'and',
 'to',
 'teach',
 'them',
 'by',
 'what',
 'name',
 'a',
 'whale',
 '-',
 'fish',
 'is',
 'to',
 'be',
 

In [4]:
with open('moby.txt', 'r') as f:
    moby_raw = f.read()

In [5]:
moby_raw

'[Moby Dick by Herman Melville 1851]\n\n\nETYMOLOGY.\n\n(Supplied by a Late Consumptive Usher to a Grammar School)\n\nThe pale Usher--threadbare in coat, heart, body, and brain; I see him\nnow.  He was ever dusting his old lexicons and grammars, with a queer\nhandkerchief, mockingly embellished with all the gay flags of all the\nknown nations of the world.  He loved to dust his old grammars; it\nsomehow mildly reminded him of his mortality.\n\n"While you take in hand to school others, and to teach them by what\nname a whale-fish is to be called in our tongue leaving out, through\nignorance, the letter H, which almost alone maketh the signification\nof the word, you deliver that which is not true." --HACKLUYT\n\n"WHALE. ... Sw. and Dan. HVAL.  This animal is named from roundness\nor rolling; for in Dan. HVALT is arched or vaulted." --WEBSTER\'S\nDICTIONARY\n\n"WHALE. ... It is more immediately from the Dut. and Ger. WALLEN;\nA.S. WALW-IAN, to roll, to wallow." --RICHARDSON\'S DICTIONARY

### Ejemplo 1

¿Cuántos tokens (palabras y símbolos de puntuación) hay en texto1?

In [7]:
len(text1)

260819

In [7]:
def ejemplo1():
    
    return len(nltk.word_tokenize(moby_raw)), len(text1), len(text1.tokens)

ejemplo1()

(255018, 260819, 260819)

In [12]:
all_tokens = nltk.word_tokenize(moby_raw)

In [15]:
# Count the frequency of each punctuation token
tokens_freq = FreqDist(all_tokens)


In [16]:
# Print tokens and frequencies in descending order
sorted_tokens_freq = sorted(tokens_freq.items(), key=lambda item: item[1], reverse=True)

In [18]:
for token, frequency in sorted_tokens_freq:
    print(f"{token}: {frequency}")


,: 19204
the: 13715
.: 7306
of: 6513
and: 6010
a: 4545
to: 4515
;: 4173
in: 3908
that: 2978
his: 2459
it: 2196
I: 2113
!: 1767
is: 1722
--: 1713
with: 1659
he: 1658
was: 1639
as: 1620
'': 1615
's: 1585
``: 1456
all: 1444
for: 1413
this: 1280
at: 1230
not: 1170
by: 1135
but: 1110
him: 1058
from: 1052
be: 1027
?: 1004
on: 1003
so: 914
one: 880
you: 841
whale: 782
had: 767
have: 763
there: 710
But: 703
or: 697
were: 679
now: 645
which: 640
me: 624
their: 612
The: 609
are: 586
they: 586
an: 582
some: 578
then: 570
my: 564
like: 558
when: 553
upon: 537
into: 519
out: 518
more: 500
up: 499
Ahab: 498
no: 483
man: 472
them: 471
ship: 454
what: 438
old: 429
ye: 427
would: 425
if: 421
been: 415
we: 411
other: 409
over: 397
these: 381
will: 377
its: 372
And: 367
sea: 365
do: 363
only: 360
down: 355
such: 336
though: 335
her: 329
any: 320
who: 318
time: 316
very: 311
It: 309
than: 309
long: 305
about: 304
said: 302
still: 299
yet: 299
those: 297
before: 293
has: 291
great: 290
must: 283
seemed: 28

In [9]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.probability import FreqDist
import string


# Tokenize the text
tokens = word_tokenize(moby_raw)

# Filter out punctuation tokens
punctuation_tokens = [token for token in tokens if token in string.punctuation]

# Count the frequency of each punctuation token
punctuation_freq = FreqDist(punctuation_tokens)

# Print the frequency distribution of punctuation tokens
print(punctuation_freq)

<FreqDist with 15 samples and 34408 outcomes>


In [11]:
for punctuation, frequency in punctuation_freq.items():
    print(f"{punctuation}: {frequency}")


[: 3
]: 3
.: 7306
(: 215
): 215
,: 19204
;: 4173
!: 1767
:: 196
': 270
-: 3
?: 1004
$: 2
*: 45
&: 2


The result <FreqDist with 15 samples and 34408 outcomes> means that there are 15 unique punctuation tokens in your text, and a total of 34,408 individual occurrences of these punctuation tokens.

To break down the interpretation further:

"15 unique samples": This refers to the number of distinct punctuation tokens present in the text. In other words, there are 15 different types of punctuation marks used.

"34408 outcomes": This represents the total count of punctuation tokens in the text. It's the sum of the occurrences of all the individual punctuation marks.

In this context, "samples" are the different types of punctuation marks (e.g., '.', ',', '!', '?', etc.), and "outcomes" are the total occurrences of these punctuation marks in the text.

If you want to see the individual punctuation marks and their corresponding frequencies, you can use the items() method of the FreqDist object to retrieve a list of tuples where each tuple contains a punctuation mark and its frequency. Here's how you can do that:

### Ejemplo 2

¿Cuántos tokens únicos (palabras únicas y símbolos de puntuación) tiene texto1?

In [8]:
def ejemplo2():
    
    return len(set(nltk.word_tokenize(moby_raw))), len(set(text1))

ejemplo2()

(20754, 19317)

### Etiquetado de palabras

In [21]:
tag = nltk.pos_tag(text1)
tag

[('[', 'JJ'),
 ('Moby', 'NNP'),
 ('Dick', 'NNP'),
 ('by', 'IN'),
 ('Herman', 'NNP'),
 ('Melville', 'NNP'),
 ('1851', 'CD'),
 (']', 'NNP'),
 ('ETYMOLOGY', 'NNP'),
 ('.', '.'),
 ('(', '('),
 ('Supplied', 'VBN'),
 ('by', 'IN'),
 ('a', 'DT'),
 ('Late', 'JJ'),
 ('Consumptive', 'NNP'),
 ('Usher', 'NNP'),
 ('to', 'TO'),
 ('a', 'DT'),
 ('Grammar', 'NNP'),
 ('School', 'NNP'),
 (')', ')'),
 ('The', 'DT'),
 ('pale', 'NN'),
 ('Usher', 'NNP'),
 ('--', ':'),
 ('threadbare', 'NN'),
 ('in', 'IN'),
 ('coat', 'NN'),
 (',', ','),
 ('heart', 'NN'),
 (',', ','),
 ('body', 'NN'),
 (',', ','),
 ('and', 'CC'),
 ('brain', 'NN'),
 (';', ':'),
 ('I', 'PRP'),
 ('see', 'VBP'),
 ('him', 'PRP'),
 ('now', 'RB'),
 ('.', '.'),
 ('He', 'PRP'),
 ('was', 'VBD'),
 ('ever', 'RB'),
 ('dusting', 'VBG'),
 ('his', 'PRP$'),
 ('old', 'JJ'),
 ('lexicons', 'NNS'),
 ('and', 'CC'),
 ('grammars', 'NNS'),
 (',', ','),
 ('with', 'IN'),
 ('a', 'DT'),
 ('queer', 'NN'),
 ('handkerchief', 'NN'),
 (',', ','),
 ('mockingly', 'RB'),
 ('embelli

### Ejemplo 3

¿Después de lematizar los verbos, cuántos tokens únicos tiene texto1?

In [36]:
from nltk.stem import WordNetLemmatizer

def example_three():
    
    lemmatizer = WordNetLemmatizer()
    lemmatized = [lemmatizer.lemmatize(t[0]) for t in tag if t[1].startswith('VB')]
    
    return len(set(lemmatized))

example_three()

5607

### Pregunta 1

¿Cuál es la diversidad léxica del texto de entrada? (ratio entre tokens únicos y el número total de tokens)

In [24]:
def respuesta1():
    respuesta =  len(set(text1)) / len(text1)
    return respuesta

respuesta1()

0.07406285585022564

### Pregunta 2

¿Qué porcentaje de tokens es 'whale'or 'Whale'?

In [25]:
def respuesta2():
    respuesta = 100 * (text1.count('whale')+text1.count('Whale')) / len(text1)
    return respuesta

respuesta2()

0.455488288813315

### Pregunta 3
¿Cuáles son los 20 tokens más frecuentes (únicos) en el texto? ¿Cuál es su frecuencia?

*Esta función debería devolver una lista de 20 tuplas donde cada tupla tiene la forma `(token, frecuencia)`. La lista debe ordenarse en orden descendente de frecuencia.*

In [26]:
from collections import Counter
def respuesta3():
    respuesta = Counter()
    for word in text1.tokens:
        respuesta[word] += 1
    return respuesta.most_common(20)

# Counter(moby_tokens).most_common(20)
''' OTRA SOLUCIÓN <tarda mucho>
    respuesta = []
    for w in text1.tokens:
        respuesta.append((w, text1.count(w)))
    respuesta = sorted(respuesta, key=lambda rpta: rpta[1])
    return respuesta[0:21]
'''
respuesta3()

[(',', 18713),
 ('the', 13721),
 ('.', 6862),
 ('of', 6536),
 ('and', 6024),
 ('a', 4569),
 ('to', 4542),
 (';', 4072),
 ('in', 3916),
 ('that', 2982),
 ("'", 2684),
 ('-', 2552),
 ('his', 2459),
 ('it', 2209),
 ('I', 2124),
 ('s', 1739),
 ('is', 1695),
 ('he', 1661),
 ('with', 1659),
 ('was', 1632)]

### Pregunta 4

¿Qué tokens tienen una longitud superior a 5 y frecuencia de más de 150?

*Esta función debería devolver una lista ordenada alfabeticamente de los tokens que coinciden con las restricciones anteriores. Para ordenar la lista, usar `sorted ()`*

In [27]:
def respuesta4():
    fdist = FreqDist(text1)
    respuesta = sorted([word for word in set(text1.tokens) if len(word) > 5 and fdist[word] > 150])
    return respuesta

''' OTRA SOLUCIÓN <tarda un poco, pero da el mismo resultado>  
    respuesta = []
    for word in set(texto1.tokens):
        if len(word) > 5 and texto1.count(word) > 150:
            respuesta.append(word)
'''

respuesta4()

['Captain',
 'Pequod',
 'Queequeg',
 'Starbuck',
 'almost',
 'before',
 'himself',
 'little',
 'seemed',
 'should',
 'though',
 'through',
 'whales',
 'without']

### Pregunta 5

Encontrar la palabra mas larga en texto1 y la longitud de dicha palabra.

*Esta función debe devolver una tupla `(palabra, longitud)`.*

In [28]:
def respuesta5():
    words_len = [(word, len(word)) for word in set(text1.tokens)]
    respuesta = sorted(words_len, key=lambda word_len: word_len[1], reverse=True)
    return respuesta[0]

respuesta5()

('uninterpenetratingly', 20)

### Pregunta 6

Qué palabras únicas tienen una frecuencia de más de 2000? ¿Cuál es su frecuencia?

"Sugerencia: es conveniente utilizar` isalpha () `para comprobar si el token es una palabra y no una puntuación".

* Esta función debe devolver una lista de tuplas de la forma `(frecuencia, palabra)` ordenadas en orden descendente de frecuencia. *

In [29]:
def respuesta6():
    fdist = FreqDist(text1)
    words_unique = [(fdist[word], word) for word in set(text1.tokens) if word.isalpha() and fdist[word] > 2000]
    respuesta = sorted(words_unique, key=lambda word: word[0], reverse=True)
    return respuesta

respuesta6()

[(13721, 'the'),
 (6536, 'of'),
 (6024, 'and'),
 (4569, 'a'),
 (4542, 'to'),
 (3916, 'in'),
 (2982, 'that'),
 (2459, 'his'),
 (2209, 'it'),
 (2124, 'I')]

### Pregunta 7
¿Cuál es el número promedio de tokens por oración?

In [30]:
import statistics
def respuesta7():
    sentences = nltk.sent_tokenize(moby_raw)
    respuesta = [len(nltk.word_tokenize(w)) for w in sentences]
    return statistics.mean(respuesta)

respuesta7()

25.88489646772229

### Pregunta 8

¿Cuáles son las 5 etiquetas gramaticales más frecuentes en el texto? ¿Cuál es su frecuencia?

* Esta función debe devolver una lista de tuplas de la forma `(etiqueta_gramatical, frecuencia)` ordenadas en orden descendente de frecuencia. *

In [32]:
def respuesta8():
    tag_fd = nltk.FreqDist(tg for (word, tg) in tag)
    respuesta = tag_fd.most_common(5)
    return respuesta
    
respuesta8()

[('NN', 36099), ('IN', 28964), ('DT', 25826), ('JJ', 18811), (',', 18713)]

In [35]:
nltk.help.upenn_tagset('NN')
nltk.help.upenn_tagset('IN')
nltk.help.upenn_tagset('DT')
nltk.help.upenn_tagset('JJ')
nltk.help.upenn_tagset(',')

NN: noun, common, singular or mass
    common-carrier cabbage knuckle-duster Casino afghan shed thermostat
    investment slide humour falloff slick wind hyena override subhumanity
    machinist ...
IN: preposition or conjunction, subordinating
    astride among uppon whether out inside pro despite on by throughout
    below within for towards near behind atop around if like until below
    next into if beside ...
DT: determiner
    all an another any both del each either every half la many much nary
    neither no some such that the them these this those
JJ: adjective or numeral, ordinal
    third ill-mannered pre-war regrettable oiled calamitous first separable
    ectoplasmic battery-powered participatory fourth still-to-be-named
    multilingual multi-disciplinary ...
,: comma
    ,
